pkg/tools/benchmark/job/service/insert.go

```diff
@@ -0,0 +1,95 @@
+//
+// Copyright (C) 2019-2023 vdaas.org vald team <vald@vdaas.org>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package service manages the main logic of benchmark job.
+package service
+
+import (
+	"context"
+	"math"
+	"strconv"
+
+	"github.com/vdaas/vald/apis/grpc/v1/payload"
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/sync/errgroup"
+)
+
+func (j *job) insert(ctx context.Context, ech chan error) error {
+	log.Info("[benchmark job] Start benchmarking insert")
+	// create data
+	vecs := j.hdf5.GetByGroupName(j.dataset.Group)
+	cfg := &payload.Insert_Config{
+		SkipStrictExistCheck: j.insertConfig.SkipStrictExistCheck,
+	}
+	if j.timestamp > int64(0) {
+		cfg.Timestamp = j.timestamp
+	}
+	eg, egctx := errgroup.New(ctx)
+	eg.SetLimit(j.concurrencyLimit)
+	for i := j.dataset.Range.Start; i <= j.dataset.Range.End; i++ {
+		iter := i
+		eg.Go(func() error {
+			log.Debugf("[benchmark job] Start insert: iter = %d", iter)
+			err := j.limiter.Wait(egctx)
+			if err != nil {
+				log.Errorf("[benchmark job] limiter error is detected: %s", err.Error())
+				if errors.Is(err, context.Canceled) {
+					return nil
+				}
+				select {
+				case <-egctx.Done():
+					return egctx.Err()
+				case ech <- err:
+				}
+			}
+			// loopCnt represents the quotient of iter divided by the len(vecs).
+			// This is to account for when iter exceeds len(vecs).
+			// It is used to calculate idx to determine which index of vecs to access.
+			// idx takes between <0, len(vecs)-1>.
+			loopCnt := math.Floor(float64(iter-1) / float64(len(vecs)))
+			idx := iter - 1 - (len(vecs) * int(loopCnt))
```

Comment:
What happnes when iter is 0?
Also can we use modulo like `iter % len(vecs)` to achieve this?

Reply:
@ykadowak 
I did not consider this situation because iter will never be `0`.
`dataset.Range.Start` which is the initial value of iter takes 1 or more.

Reply:
> dataset.Range.Start which is the initial value of iter takes 1 or more.

Is it validated somewhere?

Also what about using modulo? It looks to me you can achieve the same thing as l.62-63 with one line by using modulo. Correct me if I'm wrong.

Reply:
> > dataset.Range.Start which is the initial value of iter takes 1 or more.
> 
> Is it validated somewhere?

Yes, but only on CRD scheme. -> https://github.com/vdaas/vald/blob/feature/continuous-benchmark/charts/vald-benchmark-operator/crds/valdbenchmarkjob.yaml#L261-L272
So, there are no validated codes under the `pkg/tools/benchmark/job/`.

> Also what about using modulo? It looks to me you can achieve the same thing as l.62-63 with one line by using modulo. Correct me if I'm wrong.

The modulo will work well and make codes to be more simply!
I will try to change to use modulo.

Reply:
fixed: https://github.com/vdaas/vald/pull/2216/commits/85b9e420be22649661a6ff3b9423b3f721687bac